integration-test/common/src/test/java/com/tencent/rss/test/QuorumTest.java

```diff
@@ -0,0 +1,302 @@
+      Sets.newHashSet(shuffleServerInfo1, shuffleServerInfo2, shuffleServerInfo3),
+      testAppId, 0, 0);
+    assertEquals(report, blockIdBitmap);
+
+    // data read should success
+    Roaring64NavigableMap taskIdBitmap = Roaring64NavigableMap.bitmapOf(0);
+    SendShuffleDataResult result = shuffleWriteClientImpl.sendShuffleData(testAppId, blocks);
+    Roaring64NavigableMap failedBlockIdBitmap = Roaring64NavigableMap.bitmapOf();
+    Roaring64NavigableMap succBlockIdBitmap = Roaring64NavigableMap.bitmapOf();
+    for (Long blockId : result.getSuccessBlockIds()) {
+      succBlockIdBitmap.addLong(blockId);
+    }
+    for (Long blockId : result.getFailedBlockIds()) {
+      failedBlockIdBitmap.addLong(blockId);
+    }
+    assertEquals(0, failedBlockIdBitmap.getLongCardinality());
+    assertEquals(blockIdBitmap, succBlockIdBitmap);
+
+    ShuffleReadClientImpl readClient = new ShuffleReadClientImpl(StorageType.MEMORY_LOCALFILE.name(),
+      testAppId, 0, 0, 100, 1,
+      10, 1000, "", blockIdBitmap, taskIdBitmap,
+      Lists.newArrayList(shuffleServerInfo1, shuffleServerInfo2, shuffleServerInfo3), null);
+    validateResult(readClient, expectedData);
+
+    //case2: 2 servers are failed, the block sending should fail
+    enableTimeout((MockedShuffleServer)shuffleServers.get(1), 1000);
+    enableTimeout((MockedShuffleServer)shuffleServers.get(2), 1000);
+
+    blockIdBitmap = Roaring64NavigableMap.bitmapOf();
+    blocks = createShuffleBlockList(
+      0, 0, 0, 3, 25, blockIdBitmap,
+      expectedData, Lists.newArrayList(shuffleServerInfo1, shuffleServerInfo2, shuffleServerInfo3));
+    result = shuffleWriteClientImpl.sendShuffleData(testAppId, blocks);
+    failedBlockIdBitmap = Roaring64NavigableMap.bitmapOf();
+    succBlockIdBitmap = Roaring64NavigableMap.bitmapOf();
+    for (Long blockId : result.getSuccessBlockIds()) {
+      succBlockIdBitmap.addLong(blockId);
+    }
+    for (Long blockId : result.getFailedBlockIds()) {
+      failedBlockIdBitmap.addLong(blockId);
+    }
+    assertEquals(blockIdBitmap, failedBlockIdBitmap);
+    assertEquals(0, succBlockIdBitmap.getLongCardinality());
+
+    // report result should fail
+    partitionToBlockIds.put(0, Lists.newArrayList(blockIdBitmap.stream().iterator()));
+    try {
+      partitionToServers.put(0, Lists.newArrayList(shuffleServerInfo1, shuffleServerInfo2, shuffleServerInfo3));
+      shuffleWriteClientImpl.reportShuffleResult(partitionToServers, testAppId, 0, 0L,
+        partitionToBlockIds, 1);
+      fail();
+    } catch (Exception e){
+      assertTrue(e.getMessage().startsWith("Report shuffle result is failed"));
+    }
+    try {
+      report = shuffleWriteClientImpl.getShuffleResult("GRPC",
+        Sets.newHashSet(shuffleServerInfo1, shuffleServerInfo2, shuffleServerInfo3),
+        testAppId, 0, 0);
+      fail();
+    } catch (Exception e) {
+      assertTrue(e.getMessage().startsWith("Get shuffle result is failed"));
+    }
+
```

Comment:
case is not enough.
1. write data X, Operation of server A is successful.  Operation of server B is timeout, Operation of server C is successful. Then server c crash.
2. We also need to consider multiple data.

Reply:
What is "multiple data" ?

Reply:
> What is "multiple data" ?

many times sendata

Reply:
Add case 3 for situation1, and case 4 for situation2.

Reply:
case 3 the result should be successful.

Reply:
case 3 the result should be failed, because A/B are failed.

Reply:
> case 3 the result should be failed, because A/B are failed.

B is timeout when we write to B， When we read B, B is successful. So the case will be successful.

Reply:
The case 3 is updated

Reply:
We also need the case that shuffle server that restart.

Reply:
Add case5 to simulate when one and two server are restarted.